File: php/minesweeper/minesweeper.php
<?php

function solve(string $board){
    $g = new Game($board);
    return $g->solved;
}

class Game{
    public $board = [];
    public $solved = "";

    function __construct($board){
        $lines = explode(
            "\n",
            trim($board)
        );

        // check for unequal board
        $cl = [];
        foreach($lines as $line){
            $cl[strlen(trim($line))] = true;
        }
        if(count($cl) > 1){
            throw new InvalidArgumentException("Unequal board");
        }

        // check for valid bounds
        $bounds = [
            array_shift($lines),
            array_pop($lines)
        ];
        foreach($bounds as $bound){
            if(!preg_match('/\+-{2,}\+/', $bound)){
                throw new InvalidArgumentException("Invalid start");
            }
        }

        // check for valid line then record Point
        foreach($lines as $y => $line){
            if(!preg_match("/\|(( |\*){2,})\|/", $line, $matches)){
                throw new InvalidArgumentException("Invalid line");
            }
            foreach(str_split($matches[1]) as $x => $val){
                $this->board["$x,$y"] = new Point($this, $x, $y, $val);
            }
        }

        // solve the minesweep
        $tab = [];
        foreach($this->board as $point){
            list($x,$y) = $point->coord;
            $tab[$y][$x] = $point->count_mines();
        }

        // draw solved
        $draw = [];
        $draw[] = array_shift($bounds);
        foreach($tab as $line){
            $draw[] = "|". implode("",$line) . "|";
        }
        $draw[] = array_pop($bounds);
        $this->solved = "\n" . implode("\n",$draw) . "\n";
    }
}

class Point{
    public $coord = [];
    public $val = "";
    private $parent;

    function __construct($parent, $x, $y, $val){
        $this->parent = $parent;
        $this->coord = [$x, $y];
        $this->val = $val;
    }

    function count_mines(){
        $board = $this->parent->board;
        list($x,$y) = $this->coord;

        // if is a mine
        if($this->val === "*"){
            return "*";
        }

        $mines = 0;
        $coords = [];
        foreach(range(-1,1) as $ox){
            foreach(range(-1,1) as $oy){
                $nx = $x + $ox;
                $ny = $y + $oy;
                $coords[] = "$nx,$ny";
            }
        }

        foreach(array_unique($coords) as $coord){
            if(isset($board[$coord])){
                $point = $board[$coord];
                $mines += $point->val === "*";
            }
        }

        // return space instead of 0
        return $mines ?: " ";
    }
}

File: python/minesweeper/minesweeper.py
NEIGHBORHOOD = [(ox, oy) for ox in [-1, 0, 1]
                for oy in [-1, 0, 1] if (ox, oy) != (0, 0)]  # all around but itself


def board(input_board_array):
    # well i prefer go OOP
    g = Game(input_board_array)
    return g.solved


class Game:
    def __init__(self, board):
        self.board = dict()

        # check for Irregular board
        if len(set(len(r) for r in board)) > 1:
            raise ValueError("Irregular board")

        # fill the board
        for coord, val in (((x, y), val) for y, line in enumerate(board) for x, val in enumerate(line)):
            self.board[coord] = Point(self, coord, val)

        # solve the board
        tab = [""] * len(board)
        for coord, point in self.board.items():
            x, y = coord
            tab[y] += point.count_mines()

        self.solved = tab


class Point:
    def __init__(self, parent, coord, val):
        if val not in " *":
            raise ValueError("Irregular point")

        self.parent = parent
        self.coord = coord
        self.val = val

    def count_mines(self):
        board = self.parent.board
        x, y = self.coord

        # if is a mine
        if(self.val == "*"):
            return "*"

        mines = 0
        for ox, oy in NEIGHBORHOOD:
            ncoord = (x + ox, y + oy)
            if ncoord in board:
                mines += board[ncoord].val == "*"

        return str(mines or " ")
